pyteal/ast/opup.py

```diff
@@ -0,0 +1,163 @@
+from pyteal.ast.app import OnComplete
+from pyteal.errors import TealInputError
+from .while_ import While
+from .expr import Expr
+from .global_ import Global
+from .seq import Seq
+from .if_ import If
+from .int import Int
+from .bytes import Bytes
+from .gitxn import Gitxn
+from .itxn import InnerTxnBuilder
+from .scratch import ScratchSlot
+from .scratchvar import ScratchVar
+from .txn import TxnField, TxnType
+from .for_ import For
+from ..types import TealType, require_type
+from enum import Enum
+
+
+class OpUpMode(Enum):
+    """An Enum object that defines the mode used for the OpUp utility.
+
+    Note: the Explicit mode requires the app id to be provided
+    through the foreign apps array.
+    """
+
+    """The app to call must be provided by the user."""
+    Explicit = 0
+
+    """The app to call is created then deleted for each request to increase budget."""
+    OnCall = 1
+
+
+ON_CALL_APP = Bytes("base16", "06810143")  # v6 pyteal program "Int(1)"
+
+
+class OpUp:
+    """Utility for increasing opcode budget during app execution.
+    Note: the available budget must be high enough to execute the budget
+    increase code prior to invoking ensure_budget() or maximize_budget().
+
+    Example:
+        .. code-block:: python
+
+            opup = OpUp(OpUpMode.OnCall)
+            program_with_opup = Seq(
+                ...,
+                opup.ensure_budget(Int(1000)),
+                ...,
+            )
+    """
+
+    def __init__(self, mode: OpUpMode, target_app_id: Expr = None):
+        if mode == OpUpMode.Explicit:
+            if target_app_id is None:
+                raise TealInputError(
+                    "target_app_id must be specified in Explicit OpUp mode"
+                )
+            require_type(target_app_id, TealType.uint64)
+            self.target_app_id = target_app_id
+        elif mode == OpUpMode.OnCall:
+            self.target_app_id_slot = ScratchSlot()
+        else:
+            raise TealInputError("Invalid OpUp mode provided")
+
+        self.mode = mode
+
+        # A budget buffer is necessary to deal with an edge case of ensure_budget():
+        #   if the current budget is equal to or only slightly higher than the
+        #   required budget then it's possible for ensure_budget() to return with a
+        #   current budget less than the required budget. The buffer prevents this
+        #   from being the case.
+        self.buffer = Int(50)
+
+    def _create_app(self) -> Expr:
+        self.target_app_id = self.target_app_id_slot.load()
```

Comment:
@algoidurovic I'm not sure I follow how the statement is having the intended effect.  Can you help me understand what's happening here?

I _think_ the intent is to updated `target_app_id` for later usage in delete.  However, at the highlighted line, I don't see how the created app ID can be known.  It seems app creation happens _next_.

I'd need to inspect TEAL to be more confident, but I also _believe_ the `load` invocation needs to be part of the returned AST.

Reply:
Summarizing a verbal with @algoidurovic:

* I misinterpreted the AST evaluation.  I mistakenly treated `self.target_app_id` as a local variable.
* We discussed that it's worthwhile to consider different representations that might be less error prone to follow.  Leaving the thread open while the topic is considered though we're good to keep as is.

Reply:
I think we should have a team discussion on coding practices in PyTeal relating to this topic. There's definitely an argument for representing PyTeal "subprograms" as functions rather than variables but this may complicate the code in certain places so we likely shouldn't apply it universally even if we do adopt this practice.

for example:
```
expr = Int(1) + Int(2), program = Seq(..., expr, ...)
->
def expr():
  return Int(1) + Int(2)
program = Seq(..., expr(), ...)
```

The benefit to the second approach is that it's more clear to the developer that `expr` stores an unevaluated expression. With that said, my opinion is that this is somewhat outside the scope of this PR and shouldn't hold up approval. Let me know what you think.

Reply:
@algoidurovic A few thoughts here below.  I'll resolve because I think we're good though feel welcomed to re-open if you'd like to continue discussion.

A few thoughts:
* Thanks for giving the topic consideration.
* At least for the convention in-question, I don't feel strongly that a single convention _must_ always be followed.  I can imagine different styles making sense in different places.  Which is also to say - I think it's reasonable to try something in 1 place without applying it to _all_ places.
* I agree we need not hold up the PR.  As of https://github.com/algorand/pyteal/pull/274/commits/399a1b915f77dae95eaee41f5837e31c6d5d9ee3, the PR is in a different form and I think the concern is addressed.
* I'm up for a group discussion.  Can make for a post-stand up discussion for those interested in the topic.